staking/parse.go

```diff
@@ -3,13 +3,215 @@
+//
+// Ref: https://github.com/golang/go/blob/go1.19.12/src/crypto/x509/parser.go#L789-L968
+func ParseCertificatePermissive(bytes []byte) (*Certificate, error) {
+	if len(bytes) > MaxCertificateLen {
+		return nil, ErrCertificateTooLarge
+	}
+
+	input := cryptobyte.String(bytes)
+	// Consume the length and tag bytes.
+	if !input.ReadASN1(&input, cryptobyte_asn1.SEQUENCE) {
+		return nil, ErrMalformedCertificate
+	}
+
+	// Read the "to be signed" certificate into input.
+	if !input.ReadASN1(&input, cryptobyte_asn1.SEQUENCE) {
+		return nil, ErrMalformedTBSCertificate
+	}
+	if !input.SkipOptionalASN1(cryptobyte_asn1.Tag(0).Constructed().ContextSpecific()) {
+		return nil, ErrMalformedVersion
+	}
+	if !input.SkipASN1(cryptobyte_asn1.INTEGER) {
+		return nil, ErrMalformedSerialNumber
+	}
+	if !input.SkipASN1(cryptobyte_asn1.SEQUENCE) {
+		return nil, ErrMalformedSignatureAlgorithmIdentifier
+	}
+	if !input.SkipASN1(cryptobyte_asn1.SEQUENCE) {
+		return nil, ErrMalformedIssuer
+	}
+	if !input.SkipASN1(cryptobyte_asn1.SEQUENCE) {
+		return nil, ErrMalformedValidity
+	}
+	if !input.SkipASN1(cryptobyte_asn1.SEQUENCE) {
+		return nil, ErrMalformedIssuer
+	}
+
+	// Read the "subject public key info" into input.
+	if !input.ReadASN1(&input, cryptobyte_asn1.SEQUENCE) {
+		return nil, ErrMalformedSPKI
+	}
+
+	// Read the public key algorithm identifier.
+	var pkAISeq cryptobyte.String
+	if !input.ReadASN1(&pkAISeq, cryptobyte_asn1.SEQUENCE) {
+		return nil, ErrMalformedPublicKeyAlgorithmIdentifier
+	}
+	pkAI, err := parseAI(pkAISeq)
+	if err != nil {
+		return nil, err
+	}
+
+	// Note: Unlike the x509 package, we require parsing the public key.
+
+	var spk asn1.BitString
+	if !input.ReadASN1BitString(&spk) {
+		return nil, ErrMalformedSubjectPublicKey
+	}
+	publicKey, signatureAlgorithm, err := parsePublicKey(&publicKeyInfo{
+		Algorithm: pkAI,
+		PublicKey: spk,
+	})
+	return &Certificate{
+		Raw:                bytes,
+		SignatureAlgorithm: signatureAlgorithm,
+		PublicKey:          publicKey,
+	}, err
+}
+
+// Ref: https://github.com/golang/go/blob/go1.19.12/src/crypto/x509/parser.go#L149-L165
+func parseAI(der cryptobyte.String) (pkix.AlgorithmIdentifier, error) {
+	ai := pkix.AlgorithmIdentifier{}
+	if !der.ReadASN1ObjectIdentifier(&ai.Algorithm) {
+		return ai, ErrMalformedOID
+	}
+	if der.Empty() {
+		return ai, nil
+	}
+
+	var (
+		params cryptobyte.String
+		tag    cryptobyte_asn1.Tag
+	)
+	if !der.ReadAnyASN1Element(&params, &tag) {
+		return ai, ErrMalformedParameters
+	}
+	ai.Parameters.Tag = int(tag)
+	ai.Parameters.FullBytes = params
+	return ai, nil
+}
+
+// Ref: https://github.com/golang/go/blob/go1.19.12/src/crypto/x509/x509.go#L172-L176
+type publicKeyInfo struct {
+	Algorithm pkix.AlgorithmIdentifier
+	PublicKey asn1.BitString
+}
+
+// Ref: https://github.com/golang/go/blob/go1.19.12/src/crypto/x509/parser.go#L215-L306
+func parsePublicKey(keyData *publicKeyInfo) (any, x509.SignatureAlgorithm, error) {
```

Comment:
After Durango activates, we'll remove the `SignatureAlgorithm` from the cert - and this function.